### DISCUSSION TITLE
Problem with ref attribute use : this.$refs is undefined

### DISCUSSION THREAD

**Discussion by mowabidev** [quasarframework/quasar#11580] [General - Components / Directives / etc]
I use vue-html2pdf with quasar and I want to generate a pdf. I need to set a ref attribute for the custom component of vue-html2pdf to be able to retrieve the data from the component as below.

```
// set ref attribute
<vue-html2pdf
ref="html2Pdf"
>
//html content to generate as pdf
</vue-html2pdf >
```

```
//get ref value
function generateInvoice () {
      this.$refs.html2Pdf.generatePdf()
    };
```
But when I do this, I get an error:

> this.$refs is undefined

Can you tell me how to proceed please?

**Comment by hawkeye64** [comment]
You haven't specified if you're using Vue v2 or v3. If v3, then options API or composition API? If you need help, give more information about your environment. Usually the output of `quasar info`.

**Reply by mowabidev** [reply]
i @hawkeye64 thank you. I use quasar v2 with composition API

**Reply by hawkeye64** [reply]
In that case, you need to declare a ref for the component
```js
setup () {
  const html2Pdf = ref(null)

  //get ref value
  function generateInvoice () {
      html2Pdf.value.generatePdf()
  }

  return {
    html2Pdf,
    generateInvoice
  }
}
```

**Reply by mowabidev** [reply]
I had already tried this. But it doesn't work. It get this error from eslint `Must use '.value' to read or write the value wrapped by 'ref()'`.

But when I put .value(), the console returns `html2pdf.value is null`

**Reply by hawkeye64** [reply]
I updated my example fo missing `.value`

You still need `ref="html2Pdf"` on your component. 

Are you sure you are returning `html2Pdf`, otherwise can't hook it up properly?

**Reply by mowabidev** [reply]
Yes, I have already done so. I did exactly what you said. And it didn't work. That's why I asked the question :)

**Reply by hawkeye64** [reply]
I (and anyone else) would need to see more code. Can you create a CodePen, please?

**Reply by rstoenescu** [reply]
@hawkeye64 is spot on with his answer

**Reply by rstoenescu** [reply]
Just make sure that the component that you are using exposes the `generatePdf()` method. Check its API

**Comment by pdanpdan** [comment]
It doesn't matter if vue2 or vue3, `this.$refs` will never be available in a function that is not inside the structure of Options API.
From the look of it you are trying to access `this.$refs` in a function that is defined somewhere, so `this` is most probably `window`

```js
(function () { console.log(this) })()
=> Window {0: Window, window: Window, self: Window, document: document, name: '', location: Location, …}
```

**Reply by hawkeye64** [reply]
If I had to guess, with what's displayed above, he's using the composition API (because of `function`), in which case there is no `this`.
